%%% converge_paper.tex --- 

%% Author: wkretzsch@gmail.com
%% Version: $Id: converge_paper.tex,v 0.0 2013/05/20 09:54:17 winni Exp$

\documentclass[nobib,a4paper]{tufte-handout}

%\renewcommand{\cite}[2][0pt]{\sidenote[][#1]{\fullcite{#2}}}
%%\usepackage[debugshow,final]{graphics}

\usepackage{color}

% For comments in blue.
\newcommand{\blue}[1]{\textsf{\textbf{\textcolor{blue}{[#1]}}}}
% For comments in magenta.
\newcommand{\magenta}[1]{\textsf{\textbf{\textcolor{magenta}{[#1]}}}}
% For comments in red
\newcommand{\red}[1]{\textsf{\textbf{\textcolor{red}{[#1]}}}}
%\linespread{1.6} % for double spacing

\usepackage{graphicx}
\usepackage{booktabs}

%%\revision$Header: /Users/winni/marchini/converge/paper/converge_paper.tex,v 0.0 2013/05/20 09:54:17 winni Exp$
%%\morefloats

\newcommand{\rsq}{$\mbox{r}^2$}

% bibliography
\usepackage{csquotes}
\usepackage[backend=biber,style=authoryear-ibid, url=false,doi=false,isbn=false]{biblatex}
%\usepackage[backend=biber,style=nature]{biblatex}
\addbibresource{converge_paper.bib}

\begin{document}

%%%%##########################################################################


\section{Genotype Likelihoods}

Sequencing Reads for each of the 9300 samples were aligned to the
human reference genome GRCh37.p5 and stored in BAM format. 
The aligner used was Stampy (v1.0.17)~\autocite{Lunter2011} with
BWA (v0.5.9)~\autocite{Li2009}. 
A Base quality score recalibration table was created for each BAM file
using GATK (v2.3.9)~\autocite{McKenna2010,Trio2011}, and a
reference SNP and indel list from dbSNP version 137, excluding all
  sites after version 129. 
For each BAM file, GATKlite (v2.2.15)~\autocite{McKenna2010,Trio2011} was
used to remove the reads that were not properly mapped and recalibrate
base quality scores. This involved removing between 1--5$\%$ of reads
per sample. 

We used SNPTools~\autocite{Wang2013} to calculate genotype likelihoods (GL) at only those
SNP sites that were found to be polymorphic in the 286 Asian samples
in the 1000 Genomes Phase 1 samples~\autocite{Project2012}. \red{might want to
  address upfront why we didn't do site detection}. This resulted in
GLs at 13,837,179 SNPs. 

\section{Imputation and Phasing}

It is well established that imputation methods can substantially
improve genotype calls from low-coverage sequence data
\autocite{Project2012, Li2010, Wang2013}.
Our experience from the 1000 Genomes Projects is that the most
accurate genotype calls can be obtained by first applying the BEAGLE
method to obtain an initial set of estimated haplotypes. These
haplotypes are then used to initialise a program such as
Impute2~\autocite{Howie2009} or MaCH~\autocite{Li2010} to further refine the
estimates. 
Using this strategy on 9300 samples requires a large amount of
computational resources so we investigated more efficient methods.  

\subsection{BEAGLE experiments to reduce iterations}

\begin{margintable}[1.5in]
  \begin{tabular}{rrrr}
    \toprule
    Chr & Start & End & Sites\\
    \midrule
    20 & 7000 & 9000 & 28247\\
    20 & 35000 & 37000 & 25290\\
    20 & 57400 & 59400 & 28649\\
    \bottomrule
  \end{tabular}
  \caption{Coordinates and number of sites contained in 2mb regions
    used in BEAGLE optimization experiments.  Coordinates are given in
    kb.}
  \label{tab:2mbChunks}
\end{margintable}

Three non-overlapping 2mb regions were randomly chosen from chromosome 20
(See Table~\ref{tab:2mbChunks}). 
The GLs of 5680 CONVERGE samples and three 1000 Genomes Phase 1~\autocite{Project2012} CHB
and CHS samples sequenced on the ILLUMINA platform and downsampled to 
1x (GLs were generated using bamodel) were imputed and phased using
BEAGLE. 
Genotype calls at those sites were generated from 10x ILLUMINA
sequencing data provided by the BGI \magenta{could add more detail
  here if required}. 
Figure~\ref{fig:2mbChunks} \magenta{I should probably redo all figures for
  MAF instead of VarAF and reduce the number of experiments plotted on
  each figure as well} shows the \rsq between BEAGLE genotype dose
estimates derived from running BEAGLE for 10 and 20 iterations, and
with and without a reference panel of 496 1000 Genomes Project Phase 1~\autocite{Project2012}
Asian haplotypes. Increasing the number of iterations from 10 to 20
did not result in an increase in imputation accuracy.  

\begin{figure}[]
  \includegraphics[width=\textwidth]{{{img/rsquares.tgp.converge.82093.niter20.binned_by_varaf.1000SNP_bin_sizes.ggplot}}}
  \caption{\rsq between BEAGLE imputed genotype dose estimates and
    genotypes called from 10x ILLUMINA sequencing data as a function
    of variant allele frequency (VarAF).  BEAGLE was run with 10
    iterations with and without a reference panel and for 20
    iterations with a reference panel.  Number of sites per bin is
    shown in black.}
  \label{fig:2mbChunks}
\end{figure}


\begin{margintable}[0in]
  \begin{tabular}{rrrr}
    \toprule
    Chr & Start & End & Sites\\
    \midrule
    20 & 7650 & 8350 & 9755\\
    20 & 35650 & 36350 & 9356\\
    20 & 58050 & 58750 & 10329\\
    \bottomrule
  \end{tabular}
  \caption{Coordinates and number of sites contained in 700kb regions
    used in BEAGLE optimization experiments.  Coordinates are given in
    kb.}
  \label{tab:700kbChunks}
\end{margintable}

In order to run BEAGLE on a cluster with 2GB of RAM per core, we
investigated the effect of reducing chunk size on imputation accuracy.
From past experience of running BEAGLE on sequencing data, we estimated
that a chunk size of 700kb\sidenote[][0.3in]{while using all Asian 1000 Genome
Project Phase 1 haplotypes as a reference panel} would be sufficiently small
to fit the memory constraint.
The 700kb regions were chosen to lie centered within each 2mb chunk (see
Table~\ref{tab:700kbChunks}) to compare 2mb to 700kb regions. At the
same time, we also investigated reducing the number of iterations for
BEAGLE.   

\begin{figure}[]
  \includegraphics[width=\textwidth]{{{img/rsquares.tgp.converge.29440.700kb.binned_by_varaf.1000SNP_bin_sizes.ggplot}}}
  \caption{\rsq between BEAGLE imputed genotype dose estimates and
    genotypes called from 10x ILLUMINA sequencing data as a function
    of variant allele frequency (VarAF) for chunks of size 700kb (see
    Table~\ref{tab:700kbChunks}). BEAGLE was run for 2, 4, 6, 10, and
    20 iterations with a reference panel. Number of sites per bin is
    shown in black.}
  \label{fig:700kbChunks}
\end{figure}

Figure~\ref{fig:700kbChunks} shows imputation accuracy binned by
variant allele frequency for 2mb and 700kb regions at 2, 4, 6, 10, and
20 BEAGLE iterations. 
Although imputation accuracy for 2 and 4 iterations does decrease for
allele frequencies < 10\%, imputation accuracy for 6 and 10 iterations
are similar.


\begin{figure}[]
  \includegraphics[width=\textwidth]{{{img/rsquares.tgp.converge.29440.700kb.binned_by_location.loc1}}}
  \caption{Data from Figure~\ref{fig:700kbChunks} plotted as a function
    of chromosomal position for the first chunk of size 700kb (see
    Table~\ref{tab:700kbChunks}).}
  \label{fig:700kbChunk1}
\end{figure}

\begin{figure}[]
  \includegraphics[width=\textwidth]{{{img/rsquares.tgp.converge.29440.700kb.binned_by_location.loc2}}}
  \caption{Data from Figure~\ref{fig:700kbChunks} plotted as a function
    of chromosomal position for the second chunk of size 700kb (see
    Table~\ref{tab:700kbChunks}).}
  \label{fig:700kbChunk2}
\end{figure}

\begin{figure}[]
  \includegraphics[width=\textwidth]{{{img/rsquares.tgp.converge.29440.700kb.binned_by_location.loc3}}}
  \caption{Data from Figure~\ref{fig:700kbChunks} plotted as a function
    of chromosomal position for the third chunk of size 700kb (see
    Table~\ref{tab:700kbChunks}).}
  \label{fig:700kbChunk3}
\end{figure}

Figures~\ref{fig:700kbChunk1},~\ref{fig:700kbChunk2},~and~\ref{fig:700kbChunk3}
compare imputation accuracy as a function of genomic position of each
region, respectively. 
Although not all 700kb chunk ends see a decrease in imputation
accuracy compared to 2mb chunks, by about 50kb in from the 700kb chunk
end, any differences in imputation accuracy due to chunk size
disappear.
Accordingly, a chunk size of 700kb with 50kb overlap with neighboring
chunks, and 6 BEAGLE iterations were chosen for whole genome
imputation. 

\subsection{Imputation round 2}

BEAGLE (v3.3.2) was run on chunks containing roughly 2800 sites such
that neighboring chunks shared 400 sites. This corresponds to
approximately 700kb chunks with 50kb overlap after filtering out all
sites monomorphic in 1000 Genomes Project Phase
1~\autocite{Project2012} Asians. At each chunk, BEAGLE was run for six
iterations on the genotype likelihoods and using all 572 TGP phase 1
Asian haplotypes as a reference panel to generate genotype
probabilities.  
After imputation and phasing, the outside 200 sites of every chunk
were discarded.







%%%%##########################################################################

\printbibliography

\end{document}
